pkg/action/failback.go

```diff
@@ -458,3 +582,144 @@ func getClusterPairSchedulerConfig(clusterPairName string, namespace string) (*r
 		clientcmd.NewDefaultClientConfigLoadingRules())
 	return remoteClientConfig.ClientConfig()
 }
+
+// checkForPodStatusUsingPVC returns terminated, terminationInProgress, error
+// terminated: true if all pods using PVC are terminated
+// terminationInProgress: true if any pod using PVC is not terminated
+// error: error if any error occurs
+func checkForPodStatusUsingPVC(nsToPVCToNumberOfPods map[string]map[string]int, config *rest.Config) (bool, bool, error) {
+	coreClient, err := core.NewForConfig(config)
+	if err != nil {
+		return false, false, err
+	}
+
+	terminationInProgress := false
+	terminated := true
+	for ns, pvcToNumberOfPods := range nsToPVCToNumberOfPods {
+		for pvcName, numberOfPods := range pvcToNumberOfPods {
+			// If numberOfPods is already 0, then no need to check for pods
+			if numberOfPods > 0 {
+				pods, err := coreClient.GetPodsUsingPVC(pvcName, ns)
+				if err != nil {
+					return false, false, err
+				}
+				if len(pods) == 0 {
+					// All pods terminated for atlease one pvc
+					terminationInProgress = true
+				} else {
+					// Atleast one pvc has not been freed up
+					terminated = false
+				}
+				nsToPVCToNumberOfPods[ns][pvcName] = len(pods)
+			}
+		}
+	}
+
+	logrus.Infof("Current status of namespace to pvc to number of pods map: %v, terminated: %v, terminationInProgress: %v", nsToPVCToNumberOfPods, terminated, terminationInProgress)
+	return terminated, terminationInProgress, nil
+}
+
+// waitForVolumesBecomeFree waits for the volumes from the given namespaces from migration object to become free
+func (ac *ActionController) waitForVolumesBecomeFree(action *storkv1.Action, migration *storkv1.Migration, namespaces []string, config *rest.Config) error {
```

Comment:
I think we should have UT for this function.

Reply:
ideally yes but simulating the clients to get the desired results is becoming challenging. Will discuss offline.